pkg/storage/chunkfs/chunk.go

```diff
@@ -0,0 +1,422 @@
+	if avail+inc < size {
+		inc = ((size-avail)/files.BlockSize + 1) * files.BlockSize
+	}
+	newSize := c.mmf.Size() + inc
+	if newSize > cfg.maxChunkSize {
+		// it should be enough, because we checked the condition above
+		newSize = cfg.maxChunkSize
+	}
+
+	oldSize := c.mmf.Size()
+	if err := c.mmf.Grow(newSize); err != nil {
+		return err
+	}
+
+	if c.total == 0 {
+		// empty Chunk
+		return nil
+	}
+
+	// now, move meta to the end of the new Chunk
+	mSize := c.total * cMetaRecordSize
+	mOffset := oldSize - int64(mSize)
+	oldMBuf, err := c.mmf.Buffer(mOffset, mSize)
+	if err != nil {
+		c.logger.Errorf("could not map meta-buffer with offset %d for size=%d: %v", mOffset, mSize, err)
+		return fmt.Errorf("could not Grow Chunk: %w", fmt.Errorf("could not map meta-buffer with offset %d for size=%d: %w", mOffset, mSize, errors.ErrInternal))
+	}
+	newMBuf, err := c.mmf.Buffer(c.mmf.Size()-int64(mSize), mSize)
+	if err != nil {
+		c.logger.Errorf("could not map meta-buffer with offset %d for size=%d: %v", mOffset, mSize, err)
+		return fmt.Errorf("could not Grow Chunk: %w", fmt.Errorf("could not map meta-buffer with offset %d for size=%d: %w", mOffset, mSize, errors.ErrInternal))
+	}
+	copy(newMBuf, oldMBuf)
+
+	return nil
+}
+
+// OpenChunkReader opens new read operation. The function returns ChunkReader, which may be used for reading the chunk
+// records. The ChunkReader must be closed. The AppendRecords and Close() operations will be blocked until ALL
+// ChunkReaders are closed. So the ChunkReader should be requested for a short period of time and be closed as soon as
+// possible
+func (c *Chunk) OpenChunkReader(descending bool) (*ChunkReader, error) {
+	c.lock.RLock()
+	if c.mmf == nil {
+		// chunk is closed
+		c.lock.RUnlock()
+		return nil, fmt.Errorf("the chunk %s is closed: %w ", c.fn, errors.ErrClosed)
+	}
+	mb, err := c.getMetaBuf(c.total-1, c.total)
+	if err != nil {
+		return nil, err
+	}
+
+	cr := new(ChunkReader)
+	cr.c = c
+	cr.inc = 1
+	cr.mb = mb
+
+	if descending {
+		cr.inc = -1
+		cr.idx = cr.c.total - 1
+	}
+	return cr, nil
+}
+
+func (c *Chunk) available() int64 {
+	return c.mmf.Size() - int64(c.freeOffset+c.total*cMetaRecordSize)
+}
+
+func payloadSize(recs []*solaris.Record) int64 {
+	size := int64(0)
+	for _, r := range recs {
+		size += int64(len(r.Payload))
+	}
+	return size
+}
+
+func (cr *ChunkReader) HasNext() bool {
+	return cr.idx < cr.c.total && cr.idx > -1
+}
+
+func (cr *ChunkReader) Next() (UnsafeRecord, bool) {
+	if cr.HasNext() {
+		mr := cr.mb.get(cr.idx)
+		buf, err := cr.c.mmf.Buffer(int64(mr.offset), int(mr.size))
+		if err != nil {
+			cr.c.logger.Errorf("could not read payload for offset=%d for len=%d: %v", mr.offset, mr.size, err)
+			panic(err)
+		}
+		res := UnsafeRecord{ID: mr.ID, UnsafePayload: buf}
+		cr.idx += cr.inc
+		return res, true
+	}
+	return UnsafeRecord{}, false
+}
+
+// Close implements io.Closer
+func (cr *ChunkReader) Close() error {
+	cr.c.lock.RUnlock()
```

Comment:
q: Is it expected that calling it a few times will cause panic?

Reply:
yes. Together with wrong indexes this brutal situation should follow to panic. We can polish things here after trying out the concept, I am not sure it is good yet...